src/distributed_join.cuh

```diff
@@ -48,6 +50,53 @@ using std::chrono::high_resolution_clock;
 using std::chrono::duration_cast;
 using std::chrono::milliseconds;
 
+
+/**
+ * Communicate number of elements recieved from each rank during all-to-all communication.
+ *
+ * @param[in] send_offset Vector of length mpi_size + 1 such that send_count[i+1] - send_count[i]
+ * is the number of elements sent from the current rank to rank i during the all-to-all
+ * communication.
+ * @param[out] recv_count Vector of length mpi_size such that recv_count[i] is the number of
+ * elements received from rank i during the all-to-all communication.
+ */
+void
+communicate_sizes(
+    vector<cudf::size_type> const& offset,
+    vector<int64_t> &recv_count,
+    Communicator *communicator)
+{
+    int mpi_size = communicator->mpi_size;
+    vector<int64_t> send_count(mpi_size, -1);
+
+    for (int irank = 0; irank < mpi_size; irank++) {
+        send_count[irank] = offset[irank + 1] - offset[irank];
+    }
+
+    recv_count.resize(mpi_size, -1);
+
+    vector<MPI_Request> send_req(mpi_size);
```

Comment:
Can you explain why we use MPI here, can't we use the communicator?

Reply:
communicator is not guaranteed to be able to send/recv host buffers. I will add this note to the source code as well.